haystack/utils/http_client.py

```diff
@@ -0,0 +1,171 @@
+import os
+from typing import Optional, Callable, Any, Set, Dict, Type
+
+import requests
+from requests.adapters import HTTPAdapter
+from requests.exceptions import HTTPError
+from urllib3.util.retry import Retry
+
+from haystack.environment import HAYSTACK_REMOTE_API_TIMEOUT_SEC, HAYSTACK_REMOTE_API_MAX_RETRIES
+
+HTTP_TIMEOUT = int(os.environ.get(HAYSTACK_REMOTE_API_TIMEOUT_SEC, 30))
+HTTP_MAX_RETRIES = int(os.environ.get(HAYSTACK_REMOTE_API_MAX_RETRIES, 5))
+
+
+class HTTPClient:
```

Comment:
I find HTTPClient quite generic. Before reading this, I was thinking "Why do we need a HTTPClient instead of using requests" :smile: Now I see what's the purpose of this class and I like it. Let's call it something like `HttpSession`, or `HttpStableSession`? We should also improve the docstring to highlight the fact that we're adding sessions and retries over a simple `requests` http client.
```suggestion
class HttpSession:
```

Reply:
I don't see the need of this class to be fair, it hides tons of retry logic that in my opinion must be more explicit.
We can handle retry with `tenacity` and plain `requests`.

I also don't see the need of using `Session` what benefits does it give us? Do we need to store cookies? 

We can argue that this is useful for error mappings but that can be achieved with a mapping, switch or function that returns the error given the code. It's much easier to maintain this stuff if it's explicit.

Also why I can set the method both in `__init__` and `request`?

Let's not overcomplicate things and scrap this class, we can go with plain `request`.

Reply:
I vehemently disagree with this opinion @silvanocerza 
Please allow me to repeat myself. How we handle an increasing number of integration points with third-party REST APIs - really matters. We don't write these integration points all the time. Contributors do as well. Mistakes propagate. Our current approach is a mix of all different requests calls usages one can think of: various utils in various modules, retry tenacity and other decorators, inconsistent timeouts, etc. We're starting to see a serious downside of this in terms of maintainability and consistency. And what's even worse, our contributors are starting to replicate the use of this mess. Maybe HTTPClient/Session is not perfectly nailed but in its current incarnation is a solid remedy for the current mess.